synapseclient/core/async_utils.py

```diff
@@ -48,3 +51,77 @@ async def otel_trace_method_wrapper(self, *arg, **kwargs) -> None:
         return otel_trace_method_wrapper
 
     return decorator
+
+
+class ClassOrInstance:
+    """Helper class to allow a method to be called as a class method or instance method."""
+
+    def __init__(self, fn):
+        self.fn = fn
+
+    def __get__(self, obj, cls):
+        def f(*args, **kwds):
+            if obj is not None:
+                return self.fn(obj, *args, **kwds)
+            else:
+                return self.fn(cls, *args, **kwds)
+
+        functools.update_wrapper(f, self.fn)
+        return f
+
+
+# Adapted from
+# https://github.com/keflavich/astroquery/blob/30deafc3aa057916bcdca70733cba748f1b36b64/astroquery/utils/process_asyncs.py#L11
+def async_to_sync(cls):
+    """
+    Convert all name_of_thing_async methods to name_of_thing methods
+
+    (see
+    http://stackoverflow.com/questions/18048341/add-methods-to-a-class-generated-from-other-methods
+    for help understanding)
+    """
+
+    def create_method(async_method_name: str):
+        """Creates a replacement method for the async method."""
+
+        @ClassOrInstance
+        def newmethod(self, *args, **kwargs):
+            """The new method that will replace the non-async method."""
+
+            async def wrapper(*args, **kwargs):
+                """Wrapper for the function to be called in an async context."""
+                return await getattr(self, async_method_name)(*args, **kwargs)
+
+            try:
+                loop = asyncio.get_running_loop()
+            except RuntimeError:
+                return asyncio.run(wrapper(*args, **kwargs))
+            else:
+                nest_asyncio.apply(loop=loop)
+                return loop.run_until_complete(wrapper(*args, **kwargs))
+
+        return newmethod
+
+    methods = cls.__dict__.keys()
```

Comment:
Does this call only pull in the defined methods within a class or does it also pull in the methods that were inherited through something like class inheritance? Is that something that could lead to trouble down the line?

Reply:
I found that in the inherited mixins I needed to add the decorator to those classes as well.